### PR TITLE
Fix ComputeAtRootDomainMap with broadcast in view root domains

### LINKED ISSUE #1549: flatten(unsqueeze(tensor, -1)) does not work
### 🐛 Describe the bug

I am doing the following

```C++
TEST_F(NVFuserTest, FusionFlattenAfterUnsqueezeOutput_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  std::vector<int64_t> input_shape{512};

  TensorView* x = makeSymbolicTensor(input_shape.size(), DataType::Double);
  TensorView* bias = makeSymbolicTensor(input_shape.size(), DataType::Double);
  fusion.addInput(x);
  fusion.addInput(bias);

  auto x_add_bias = add(x, bias);
  auto x_unsqueeze = unsqueeze(x_add_bias, -1);
  auto x_view = flatten(x_unsqueeze);
  fusion.addOutput(x_view);

  auto options = at::TensorOptions().dtype(at::kDouble).device(at::kCUDA, 0);
  at::Tensor at_x = at::randn(input_shape, options);
  at::Tensor at_bias = at::randn(input_shape, options);
  std::vector<IValue> aten_inputs = {at_x, at_bias};

  x_view->split(0, 4);
  x_add_bias->computeAt(x_view, 1);
  x_view->axis(0)->parallelize(ParallelType::TIDx);

  FusionExecutor fe;
  fe.compileFusion(&fusion, aten_inputs);
  auto outputs = fe.runFusion(aten_inputs);

  auto at_x_add_bias = at_x + at_bias;
  auto at_x_view = at_x_add_bias.unsqueeze(-1).flatten();

  testValidate(&fusion, outputs, aten_inputs, {at_x_view}, __LINE__, __FILE__);
}
```

And getting an error

```
C++ exception with description "it != bcast_map_.end() INTERNAL ASSERT FAILED at "/home/gaoxiang/pytorch-nvfuser/torch/csrc/jit/codegen/cuda/root_domain_map.cpp":494, please report a bug to PyTorch. Not found: {[ iS8{( ceilDiv(( i0 * 1 ), 4) )}rf, iS9{4}rf ] (root: iS5{i0} bS6{1}, maybe rfactor: iS7{( i0 * 1 )}rf), bS6{1}}
Exception raised from getConcretizedKeys at /home/gaoxiang/pytorch-nvfuser/torch/csrc/jit/codegen/cuda/root_domain_map.cpp:494 (most recent call first):
frame #0: <unknown function> + 0x59b6f (0x7f559872eb6f in /home/gaoxiang/pytorch-nvfuser/build/lib/libc10.so)
frame #1: <unknown function> + 0x598d5 (0x7f559872e8d5 in /home/gaoxiang/pytorch-nvfuser/build/lib/libc10.so)
frame #2: <unknown function> + 0x59699 (0x7f559872e699 in /home/gaoxiang/pytorch-nvfuser/build/lib/libc10.so)
frame #3: <unknown function> + 0x5a728 (0x7f559872f728 in /home/gaoxiang/pytorch-nvfuser/build/lib/libc10.so)
frame #4: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f559872d544 in /home/gaoxiang/pytorch-nvfuser/build/lib/libc10.so)
frame #5: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x7b (0x7f559872b274 in /home/gaoxiang/pytorch-nvfuser/build/lib/libc10.so)
frame #6: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x59 (0x7f559872b495 in /home/gaoxiang/pytorch-nvfuser/build/lib/libc10.so)
frame #7: torch::jit::fuser::cuda::ComputeAtRootDomainMap::getConcretizedKeys(torch::jit::fuser::cuda::TensorDomain const*, torch::jit::fuser::cuda::IterDomain const*) const + 0x127 (0x7f5566d4cc51 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #8: torch::jit::fuser::cuda::ComputeAtRootDomainMapBuilder::setMaybeMapped(torch::jit::fuser::cuda::TensorDomain const*, torch::jit::fuser::cuda::IterDomain const*, torch::jit::fuser::cuda::TensorDomain const*, torch::jit::fuser::cuda::IterDomain const*) + 0x1b4 (0x7f5566d4eaa6 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #9: torch::jit::fuser::cuda::ComputeAtRootDomainMapBuilder::mapPointwiseOrReductionOp(torch::jit::fuser::cuda::Expr*) + 0x574 (0x7f5566d4f404 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #10: <unknown function> + 0x5b4cd53 (0x7f5566d51d53 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #11: void torch::jit::fuser::cuda::Expr::dispatch<torch::jit::fuser::cuda::OptOutDispatch*>(torch::jit::fuser::cuda::OptOutDispatch*, torch::jit::fuser::cuda::Expr*) + 0x355 (0x7f5566a62169 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #12: torch::jit::fuser::cuda::OptOutDispatch::handle(torch::jit::fuser::cuda::Expr*) + 0x23 (0x7f5566a5f7e3 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #13: torch::jit::fuser::cuda::BackwardVisitor::handle(torch::jit::fuser::cuda::Expr*) + 0x23 (0x7f5566bd8635 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #14: torch::jit::fuser::cuda::ComputeAtRootDomainMapBuilder::handle(torch::jit::fuser::cuda::Expr*) + 0x7d (0x7f5566d4ee5d in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #15: void torch::jit::fuser::cuda::Statement::dispatch<torch::jit::fuser::cuda::OptOutDispatch*>(torch::jit::fuser::cuda::OptOutDispatch*, torch::jit::fuser::cuda::Statement*) + 0x91 (0x7f5566a61195 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #16: torch::jit::fuser::cuda::OptOutDispatch::handle(torch::jit::fuser::cuda::Statement*) + 0x23 (0x7f5566a5f7bd in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #17: torch::jit::fuser::cuda::BackwardVisitor::handle(torch::jit::fuser::cuda::Statement*) + 0x23 (0x7f5566bd860f in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #18: torch::jit::fuser::cuda::BackwardVisitor::traverseFrom(torch::jit::fuser::cuda::Fusion*, std::vector<torch::jit::fuser::cuda::Val*, std::allocator<torch::jit::fuser::cuda::Val*> > const&, bool) + 0x6d4 (0x7f5566bd8d32 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #19: torch::jit::fuser::cuda::ComputeAtRootDomainMapBuilder::ComputeAtRootDomainMapBuilder(torch::jit::fuser::cuda::ComputeAtRootDomainMap&, bool) + 0x153 (0x7f5566d4d985 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #20: torch::jit::fuser::cuda::ComputeAtRootDomainMap::build(bool) + 0x84 (0x7f5566d4bc72 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #21: <unknown function> + 0x5808669 (0x7f5566a0d669 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #22: <unknown function> + 0x58050b5 (0x7f5566a0a0b5 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #23: torch::jit::fuser::cuda::TensorView::computeAt(torch::jit::fuser::cuda::TensorView*, int, torch::jit::fuser::cuda::ComputeAtMode) + 0x20f (0x7f5566e032e3 in /home/gaoxiang/pytorch-nvfuser/build/lib/libtorch_cuda.so)
frame #24: torch::jit::NVFuserTest_FusionFlattenAfterUnsqueezeOutput_CUDA_Test::TestBody() + 0x3ef (0x562f8c22c2e5 in ./build/bin/test_jit)
frame #25: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x65 (0x562f8c27bedb in ./build/bin/test_jit)
frame #26: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x5a (0x562f8c2756a7 in ./build/bin/test_jit)
frame #27: testing::Test::Run() + 0x100 (0x562f8c24cec2 in ./build/bin/test_jit)
frame #28: testing::TestInfo::Run() + 0x142 (0x562f8c24d8e6 in ./build/bin/test_jit)
frame #29: testing::TestSuite::Run() + 0x139 (0x562f8c24e199 in ./build/bin/test_jit)
frame #30: testing::internal::UnitTestImpl::RunAllTests() + 0x3f1 (0x562f8c25d74f in ./build/bin/test_jit)
frame #31: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x65 (0x562f8c27d14b in ./build/bin/test_jit)
frame #32: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x5a (0x562f8c27662d in ./build/bin/test_jit)
frame #33: testing::UnitTest::Run() + 0xc3 (0x562f8c25bf8f in ./build/bin/test_jit)
frame #34: <unknown function> + 0x206a45 (0x562f8bc4ea45 in ./build/bin/test_jit)
frame #35: main + 0x19c (0x562f8bc4e95a in ./build/bin/test_jit)
frame #36: <unknown function> + 0x2d310 (0x7f5560cdc310 in /usr/lib/libc.so.6)
frame #37: __libc_start_main + 0x81 (0x7f5560cdc3c1 in /usr/lib/libc.so.6)
frame #38: _start + 0x25 (0x562f8bc4e645 in ./build/bin/test_jit)
" thrown in the test body.
```

Looks like this is because merging broadcasting dim with a non-broadcasting dim is not well supported yet? I have a similar pattern in https://github.com/csarofeen/pytorch/pull/1512, and I am struggling a lot on how to make the transform replay work if I flatten the vector split ID with a normal ID.

### Versions

devel

### PR DESCRIPTION
**Pull request by naoyam**
When a broadcast root domain is merged with a non-broadcast root domain in a view transformation, it essentially means it disappears without concretization. The concretization map needs to be initialized for such domains.

This should fix #1549, but I think there's a bigger problem in `ComputeAtRootDomainMap`: #1551.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Would you mind adding the test in https://github.com/csarofeen/pytorch/issues/1549?

**Comment by naoyam** [comment]
> Would you mind adding the test in #1549?

Done

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 1]
```diff
@@ -936,6 +940,13 @@ void ComputeAtRootDomainMapBuilder::handle(TensorView* tv) {
       mapAllConsumers(DomainKey(td, id));
     }
   }
+  if (td->hasRFactor()) {
+    for (auto id : TensorDomain::noReductions(td->getRootDomain())) {
+      if (id->isBroadcast()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp:949]: @rdspring1 I thought broadcast relative to view was always defined in a following expression.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 2]
```diff
@@ -936,6 +940,13 @@ void ComputeAtRootDomainMapBuilder::handle(TensorView* tv) {
       mapAllConsumers(DomainKey(td, id));
     }
   }
+  if (td->hasRFactor()) {
+    for (auto id : TensorDomain::noReductions(td->getRootDomain())) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp:948]: Can you pull: https://github.com/csarofeen/pytorch/pull/1548/files#diff-296b2445f5a818c3240b26185b37b309fa554f56b909f73109917cbdfd580333R1375-R1390 into this PR?

**Review by csarofeen — APPROVED** [review]
Doesn't seem harmful, though curious why it's needed. Stamping since it fixes an issue and doesn't seem like it would be an otherwise negative change.

**Comment by naoyam** [comment]
~Previously broadcast domains were always created at root domains, and so `ComputeAtRootDomainMapBuilder` initializes the broadcast conretization map by looking the root domain of each TensorView. Now with `view` expressions, broadcasts can appear at an rfactor domain without appearing at its root domain, so initialization of the map for the broadcast domain can be missed. This PR makes sure those broadcast domains are always initialized.~

Sorry, that was not correct at all. The problem was that a broadcast domain of an input to `view` is retained in the root domain of the output of the `view` expression but can disappear at its rfactor domain. In that case, since the initialization of the conretization map only looks at the rfactor domain, it misses the initialization of the mapping for the broadcast domain. Added a comment to the code.

**Comment by csarofeen** [comment]
Yes, I thought that was the issue you experienced. I just thought we avoided merging broadcast dimensions as the view operation, and instead trivially reduced them, then rebroadcasted after the view operation.

**Comment by rdspring1** [comment]
`View` doesn't merge broadcast domains in its rfactor domain, so it sidesteps this issue. However, I think with `flatten` we needed to merge broadcast domains.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 1 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp:949]: Yes, we apply broadcast after the view operation.

**Pull request merged**